lib/referencer.js

```diff
@@ -272,7 +272,8 @@ class Referencer extends esrecurse.Visitor {
                 ));
         }
 
-        this.visit(node.superClass);
+		let mayRefClassId = node.superClass && node.superClass.type !== 'Identifier'
```

Comment:
I still don't see why Identifier would be a special case. Per the specification ([15.7.14 Runtime Semantics: ClassDefinitionEvaluation](https://tc39.es/ecma262/#sec-runtime-semantics-classdefinitionevaluation)), ClassHeritage is always evaluated in the class scope.

Here's an example that demonstrates this:

```js
class C {}

(class C extends C {}); // ReferenceError: Cannot access 'C' before initialization
```

Reply:
The example you provided is indeed correct, and I understand your point. However, you might be referring to the scope during dynamic runtime: In this code, 'class C extends C' throws an error because defining a class C that extends C is semantically contradictory, hence the error. However, eslint-scope is a concept of static scope. In 'class C extends C', the second C should obviously be the C in the scope where this code is located. That's why I proposed the solution above: when the target of 'extends' is an Identifier, we need to traverse this node in the current scope first, then create a class sub-scope to traverse other nodes.

[other information](https://github.com/eslint/eslint-scope/pull/113#discussion_r1428195841)

Reply:
> However, eslint-scope is a concept of static scope. In `class C extends C`, the second C should obviously be the C in the scope where this code is located.

eslint-scope aims to represent JavaScript scoping. If the second C is, per the specification, reference to the first C, it should be the same in eslint-scope. If the existing tests are wrong, we can fix them.

Reply:
I was just looking because I was curious - `@typescript-eslint/scope-manager` works fine for this case.

The fix we implemented was simply to always visit the superclass after the class name is defined. Not conditionally - just always do it.

https://github.com/typescript-eslint/typescript-eslint/blob/6128a02cb15d500fe22fe265c83e4d7a73ae52c3/packages/scope-manager/src/referencer/ClassVisitor.ts#L51-L72